File: anny_dang/homework2/singlyLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def insertAtFront(head, val):
    new_head = Node(val)
    new_head.next = head
    return new_head

def insertAtBack(head, val):
    if not head:
        return Node(val)
    
    cur = head 
    while cur.next:
        cur = cur.next
    cur.next = Node(val)

    return head

def insertAfter(head, val, loc):
    """
    assume that loc always exists in head
    """
    new_node = Node(val)
    temp = loc.next
    loc.next = new_node
    new_node.next = temp
    return head

def insertBefore(head, val, loc):
    """
    assume that loc always exists in head
    """
    if loc == head:
        return insertAtFront(head, val)
    
    prev = None
    cur = head
    while cur and cur != loc:
        prev = cur
        cur = cur.next

    new_node = Node(val)
    prev.next = new_node
    new_node.next = loc
    return head

def deleteFront(head):
    if not head:
        return head
    
    head = head.next
    return head 

def deleteBack(head):
    if not head:
        return head
    if not head.next:
        return None
    
    prev = None
    cur = head
    while cur and cur.next:
        prev = cur
        cur = cur.next
    
    prev.next = None
    return head

def deleteNode(head, loc):
    """
    assume head exists and loc always exists in head
    """
    if head == loc:
        return deleteFront(head)
    
    prev = None
    cur = head
    while cur and cur != loc:
        prev = cur
        cur = cur.next

    temp = loc.next
    prev.next = temp
    return head

def length(head):
    l = 0
    cur = head
    while cur:
        cur = cur.next
        l += 1

    return l 

def reverseIterative(head):
    prev = None
    cur = head
    while cur:
        nxt = cur.next
        cur.next = prev
        prev = cur
        cur = nxt
    
    return prev

def reverseRecursive(head):
    def recursion(prev, cur):
        if not cur:
            return prev
        
        nxt = cur.next
        cur.next = prev
        return recursion(cur, nxt)

    return recursion(None, head) 

File: homework2/rodrigo_canto/SinglyLinkedList.cpp
#include "bits/stdc++.h"
using namespace std;

struct Node {
    int data;
    Node* next;
};

Node* createNode(int val){

    Node* node = new Node;
    node -> data = val;
    node -> next = nullptr;
    return node;
}

Node* insertAtFront(Node* head, int val){

    if(head == nullptr){
        head = createNode(0);
        head -> next = createNode(val);
        return head;
    }

    Node *first = head -> next;
    Node *new_node = createNode(val);

    head -> next = new_node;
    new_node -> next = first;

    return head;
}

Node* insertAtBack(Node* head, int val){

    if(head == nullptr){
        head = createNode(0);
        head -> next = createNode(val);
        return head;
    }

    Node *curr = head;

    while(curr -> next != nullptr){
        curr = curr -> next;
    }

    Node* new_node = createNode(val);

    curr -> next = new_node;

    return head;
}

Node* insertAfter(Node* head, int val, Node* loc){

    Node* sig = loc -> next;

    Node* new_node = createNode(val);

    loc -> next = new_node;
    new_node -> next = sig;

    return head;
}

Node* insertBefore(Node* head, int val, Node* loc){

    Node* curr = head;

    while(curr -> next != loc) {
        curr = curr -> next;
    }

    head = insertAfter(head, val, curr);

    return head;
}

Node* deleteFront(Node* head){

    if(head == nullptr) return head;

    Node* first = head -> next;
    Node* second = first -> next;

    head -> next = second;

    if(head -> next == nullptr) head = nullptr;
    return head;
}

Node* deleteBack(Node* head){

    if(head == nullptr) return head;

    Node* curr = head;

    while(curr -> next -> next != nullptr){
        curr = curr -> next;
    }

    curr -> next = nullptr;

    if(head -> next == nullptr) head = nullptr;
    return head;
}

Node* deleteNode(Node* head, Node* loc){

    if(head == nullptr) return head;

    if(head -> next == loc){
        
        head -> next = head -> next -> next;
        if(head -> next == nullptr) head = nullptr;
        return head;
    }
    else{

        Node* prev = head -> next;
        Node* curr = prev -> next;

        while(curr != loc){
        
            prev = curr;
            curr = curr -> next;
        }

        prev -> next = curr -> next;
        return head;
    }
}

int length(Node* head){

    if(head == nullptr) return 0;

    int len = 0;

    Node* curr = head;

    while(curr -> next != nullptr){
        len++;
        curr = curr -> next;
    }

    return len;
}

Node* reverseIterative(Node* head){

    if(head == nullptr) return head;

    Node* ant = head -> next;
    Node* curr = ant -> next;

    ant -> next = nullptr;
    Node* sig = nullptr;

    while(curr != nullptr){

        sig = curr -> next;
        curr -> next = ant;

        ant = curr;
        curr = sig;
    }

    head -> next = ant;

    return head;
}

Node* reverseLinkedList(Node* head, Node* curr, Node* &prev){

    if(curr -> next == nullptr){
        head -> next = curr;
        prev = curr;
        return head;
    }

    head = reverseLinkedList(head, curr -> next, prev);

    prev -> next = curr;
    prev = curr;

    return head;
}

Node* reverseRecursive(Node* head){

    if(head == nullptr) return head;

    if(head -> next -> next == nullptr) return head;

    Node* prev = nullptr;
    
    head = reverseLinkedList(head, head -> next, prev);

    prev -> next = nullptr;

    return head;
}

void printLinkedList(Node* head){

    if(head == nullptr) return;

    Node* curr = head -> next;

    while(curr != nullptr){
        cout << curr -> data << " ";
        curr = curr -> next;
    }

    cout << "\n";
}

int main(){

    //Just testing functions.

    vector<int> arr = {1, 2, 3, 4, 5, 6};

    Node* head = nullptr;
    
    for(auto val : arr){
        head = insertAtBack(head, val);
    }

    printLinkedList(head);
}

